test/common/network/listener_impl_test.cc

```diff
@@ -38,4 +39,60 @@ TEST(ListenerImplDeathTest, ErrorCallback) {
   EXPECT_DEATH(errorCallbackTest(), ".*listener accept failure.*");
 }
 
+class TestListenerImpl : public ListenerImpl {
+public:
+  TestListenerImpl(Network::ConnectionHandler& conn_handler, Event::DispatcherImpl& dispatcher,
+                   ListenSocket& socket, ListenerCallbacks& cb, Stats::Store& stats_store,
+                   bool bind_to_port, bool use_proxy_proto, bool use_orig_dst)
+      : ListenerImpl(conn_handler, dispatcher, socket, cb, stats_store, bind_to_port,
+                     use_proxy_proto, use_orig_dst) {}
+  ~TestListenerImpl() {}
+
+  MOCK_METHOD1(getAddressPort_, uint16_t(sockaddr*));
+  MOCK_METHOD2(newConnection_, void(int, sockaddr*));
+
+  void newConnection(int fd, sockaddr* addr) override { newConnection_(fd, addr); }
+  void newConnection(int fd, const std::string& addr) override {
+    ListenerImpl::newConnection(fd, addr);
+  }
+
+protected:
+  uint16_t getAddressPort(sockaddr* sock) override { return getAddressPort_(sock); }
+};
+
+static void useOriginalDst() {
+  Stats::IsolatedStoreImpl stats_store;
+  Event::DispatcherImpl dispatcher;
+  Network::TcpListenSocket socket(uint32_t(10000), true);
+  Network::TcpListenSocket socketDst(uint32_t(10001), false);
+  Network::MockListenerCallbacks listener_callbacks;
+  Network::MockConnectionHandler connection_handler;
+  Network::TestListenerImpl listener(connection_handler, dispatcher, socket, listener_callbacks,
+                                     stats_store, true, false, true);
+  Network::TestListenerImpl listenerDst(connection_handler, dispatcher, socketDst,
+                                        listener_callbacks, stats_store, false, false, false);
+
+  Network::ClientConnectionPtr client_connection =
+      dispatcher.createClientConnection("tcp://127.0.0.1:10000");
+  client_connection->connect();
+
+  EXPECT_CALL(listener, getAddressPort_(_)).WillRepeatedly(Return(10001));
+  EXPECT_CALL(connection_handler, findListener("10001")).WillRepeatedly(Return(&listenerDst));
+
+  EXPECT_CALL(listener, newConnection_(_, _)).Times(0);
+  EXPECT_CALL(listenerDst, newConnection_(_, _))
+      .Times(1)
+      .WillOnce(Invoke([&](int, sockaddr*) -> void {
+        client_connection->close(ConnectionCloseType::NoFlush);
+        socket.close();
+        socketDst.close();
```

Comment:
These 2 lines in the above test are just kind of a hack to force the error callback to be called. (Sorry I should have a comment to that effect, perhaps you could add one).

Can we make this test a non-death test? If you just call dispatcher.exit() here the event loop will exit and the test will do what is expected I think (expect a call of newConnection on the other listener).